controller/health.go

```diff
@@ -72,6 +86,13 @@ func setApplicationHealth(resources []managedResource, statuses []appv1.Resource
 
 		if health.IsWorse(appHealth.Status, healthStatus.Status) {
 			appHealth.Status = healthStatus.Status
+			if persistResourceHealth {
+				appHealth.LastTransitionTime = statuses[i].Health.LastTransitionTime
+			} else {
+				appHealth.LastTransitionTime = now
+			}
+		} else if healthStatus.Status == health.HealthStatusHealthy {
+			appHealth.LastTransitionTime = lastTransitionTime
```

Comment:
I've managed to repro the issue. Unfortunately it persists with your suggestion. I'll look into it deeper tomorrow

Reply:
I've added another test to `appcontroller_test.go` where the issue you're describing doesn't happen. I think the code works as expected, but maybe you can have another look :)

Reply:
I ran a manual test by creating a pod with delayed start. The time is updated for `Missing->Progressing` state change but it doesn't get updated when state moves from `Progressing->Healthy`.

https://github.com/argoproj/argo-cd/assets/61187612/bb2ea680-1a19-4aec-a343-3facad6abd01

Reply:
That's interesting. Let me have a closer look

Reply:
I've added some more extensive testing. Currently they break because I couldn't find a way to make them work with the way we're using mocks. Any suggestions on how to make `TestUpdateHealthStatusProgression` work are welcome

Reply:
@mkieweg I added some changes which should fix your tests

Reply:
@hiddeco [wrote](https://github.com/argoproj/argo-cd/issues/16972#issuecomment-2190044843) in #16972 that this doesn't quite account for all edge cases.

By modifying @svghadi's suggestion slightly I think that would be achievable by doing this:

```diff
diff --git a/controller/health.go b/controller/health.go
index b1c457776..e1a2c5d19 100644
--- a/controller/health.go
+++ b/controller/health.go
@@ -86,15 +86,11 @@ func setApplicationHealth(resources []managedResource, statuses []appv1.Resource
 
 		if health.IsWorse(appHealth.Status, healthStatus.Status) {
 			appHealth.Status = healthStatus.Status
-			if persistResourceHealth {
-				appHealth.LastTransitionTime = statuses[i].Health.LastTransitionTime
-			} else {
-				appHealth.LastTransitionTime = now
-			}
-		} else if healthStatus.Status == health.HealthStatusHealthy {
-			appHealth.LastTransitionTime = lastTransitionTime
 		}
 	}

 	if persistResourceHealth {
+          if app.Status.Health.Status == appHealth.Status && time.Since(lastTransitionTime.Time) < 30*time.Second {
+		  appHealth.LastTransitionTime = lastTransitionTime
+          } else {
+                 appHealth.LastTransitionTime = metav1.Now()
+          }
 		app.Status.ResourceHealthSource = appv1.ResourceHealthLocationInline
 	} else {
```

Reply:
My suggestion however would be to have the heartbeat timestamp as a separate field (as e.g. "last observed" or "heartbeat"), as otherwise it could create a false impression that a transition did take place.